Mathlib/Algebra/Category/ModuleCat/OverFilteredColimits.lean

```diff
@@ -0,0 +1,351 @@
+
+protected lemma one_smul :
+    hsmul i₁ i₂ (1 : ℛ.obj c₁) m = (ℳ.map i₂ m) := by
+  simp [hsmul]
+
+protected lemma mul_smul (r₁ r₂ : ℛ.obj c₁) : hsmul i₁ i₂ (r₁ * r₂) m =
+    hsmul i₁ (𝟙 _) r₁ (hsmul i₁ i₂ r₂ m) := by
+  simp only [hsmul, map_mul, mul_smul]
+  erw [ℳ.map_id, id_apply]
+
+protected lemma smul_zero : hsmul (ℳ := ℳ) i₁ i₂ r 0 = 0 := by
+  simp [hsmul]
+
+protected lemma smul_add (m₁ m₂ : ℳ.obj c₂) : hsmul i₁ i₂ r (m₁ + m₂) =
+    hsmul i₁ i₂ r m₁ + hsmul i₁ i₂ r m₂ := by
+  simp [hsmul, smul_add]
+
+protected lemma add_smul (r₁ r₂ : ℛ.obj c₁) (m : ℳ.obj c₂) :
+    hsmul i₁ i₂ (r₁ + r₂) m = hsmul i₁ i₂ r₁ m + hsmul i₁ i₂ r₂ m := by
+  simp [hsmul, add_smul]
+
+protected lemma zero_smul : hsmul i₁ i₂ (0 : ℛ.obj c₁) m = 0 := by
+  simp [hsmul]
+
+set_option maxHeartbeats 800000 in
+lemma respect_ι
+    {c₁ c₂ c₃ : J} (i₁ : c₁ ⟶ c₃) (i₂ : c₂ ⟶ c₃)
+    (r : ℛ.obj c₁) (x : ℳ.obj c₂)
+    {d₁ d₂ d₃ : J} (j₁ : d₁ ⟶ d₃) (j₂ :  d₂ ⟶ d₃)
+    (r' : ℛ.obj d₁) (x' : ℳ.obj d₂)
+    (hrr' : colimit.ι ℛ _ r = colimit.ι ℛ _ r')
+    (hmm' : colimit.ι ℳ _ x = colimit.ι ℳ _ x') :
+    colimit.ι ℳ _ (hsmul i₁ i₂ r x) =
+    colimit.ι ℳ _ (hsmul j₁ j₂ r' x') := by
+  classical
+  obtain ⟨m, fm₁, fm₂, hm⟩ := Concrete.colimit_exists_of_rep_eq (h := hrr')
+  obtain ⟨n, fn₁, fn₂, hn⟩ := Concrete.colimit_exists_of_rep_eq (h := hmm')
+  erw [Concrete.colimit_rep_eq_iff_exists]
+  delta hsmul
+  let O : Finset J := { c₁, c₂, c₃, d₁, d₂, d₃, m, n }
+  let H : Finset ((X : J) ×' (Y : J) ×' (_ : X ∈ O) ×' (_ : Y ∈ O) ×' (X ⟶ Y)) :=
+  { ⟨c₁, m, by simp [O], by simp [O], fm₁⟩,
+    ⟨d₁, m, by simp [O], by simp [O], fm₂⟩,
+    ⟨c₂, n, by simp [O], by simp [O], fn₁⟩,
+    ⟨d₂, n, by simp [O], by simp [O], fn₂⟩,
+    ⟨c₁, c₃, by simp [O], by simp [O], i₁⟩,
+    ⟨c₂, c₃, by simp [O], by simp [O], i₂⟩,
+    ⟨d₁, d₃, by simp [O], by simp [O], j₁⟩,
+    ⟨d₂, d₃, by simp [O], by simp [O], j₂⟩ }
+
+  let S := IsFiltered.sup O H
+
+  refine ⟨S, IsFiltered.toSup O H (by simp [O]), IsFiltered.toSup _ _ (by simp [O]), ?_⟩
+  erw [compatible_smul.out, compatible_smul.out]
+  apply_fun ℛ.map (IsFiltered.toSup O H (by simp [O])) at hm
+  rw [← comp_apply, ← comp_apply, ← ℛ.map_comp, ← ℛ.map_comp] at hm
+
+  apply_fun ℳ.map (IsFiltered.toSup O H (by simp [O])) at hn
+  rw [← comp_apply, ← comp_apply, ← ℳ.map_comp, ← ℳ.map_comp] at hn
+
+  rw [← comp_apply, ← comp_apply, ← comp_apply, ← comp_apply, ← ℛ.map_comp, ← ℛ.map_comp,
+    ← ℳ.map_comp, ← ℳ.map_comp]
+  convert congr($hm • $hn) using 1 <;> congr 3
+  · erw [IsFiltered.toSup_commutes O H (f := i₁), IsFiltered.toSup_commutes O H (f := fm₁)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := i₂), IsFiltered.toSup_commutes O H (f := fn₁)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := j₁), IsFiltered.toSup_commutes O H (f := fm₂)] <;>
+    simp [O, H]
+  · erw [IsFiltered.toSup_commutes O H (f := j₂), IsFiltered.toSup_commutes O H (f := fn₂)] <;>
+    simp [O, H]
```

Comment:
This is just too costly, it seems. If you squeeze them then the proof is about 5 times faster and you don't need the heartbeats bump.
```suggestion
  · erw [IsFiltered.toSup_commutes O H (f := i₁), IsFiltered.toSup_commutes O H (f := fm₁)]
    · simp only [Finset.mem_insert, Finset.mem_singleton, true_or, O]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, O, H]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, or_true, O, H]
  · erw [IsFiltered.toSup_commutes O H (f := i₂), IsFiltered.toSup_commutes O H (f := fn₁)]
    · simp only [Finset.mem_insert, Finset.mem_singleton, true_or, or_true, O]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, or_true, O, H]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, or_true, O, H]
  · erw [IsFiltered.toSup_commutes O H (f := j₁), IsFiltered.toSup_commutes O H (f := fm₂)]
    · simp only [Finset.mem_insert, Finset.mem_singleton, true_or, or_true, O]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, or_true, O, H]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      true_or, or_true, O, H]
  · erw [IsFiltered.toSup_commutes O H (f := j₂), IsFiltered.toSup_commutes O H (f := fn₂)]
    · simp only [Finset.mem_insert, Finset.mem_singleton, true_or, or_true, O]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, Finset.mem_singleton, heq_eq_eq, true_and,
      true_or, or_true, O, H]
    · simp only [Finset.mem_insert, PSigma.mk.injEq, heq_eq_eq, true_and, Finset.mem_singleton,
      or_true, O, H]
```